proposals/safe-fixed-size-buffers.md

```diff
@@ -0,0 +1,399 @@
+
+### Metadata
+
+#### Metadata emit and code generation
+
+For metadata encoding compiler will rely on recently added [```System.Runtime.CompilerServices.InlineArrayAttribute```](https://github.com/dotnet/runtime/issues/61135). 
+
+Fixed-size buffers like:
+``` C#
+public partial class C
+{
+    public int buffer1[10];
+    public readonly int buffer2[10];
+}
+```
+will be emitted as  fields of a specially decorated struct type.
+
+Equivalent C# code will be:
+
+``` C#
+public partial class C
+{
+    public Buffer10<int> buffer1;
+    public readonly Buffer10<int> buffer2;
+}
+
+[System.Runtime.CompilerServices.InlineArray(10)]
+public struct Buffer10<T>
+{
+    private T _element0;
+
+    [UnscopedRef]
+    public System.Span<T> AsSpan()
+    {
+        return System.Runtime.InteropServices.MemoryMarshal.CreateSpan(ref _element0, 10);
+    }
+
+    [UnscopedRef]
+    public readonly System.ReadOnlySpan<T> AsReadOnlySpan()
+    {
+        // Note, an attempt to compile reports error CS1605: Cannot use '_element0' as a ref or out value because it is read-only
+        return System.Runtime.InteropServices.MemoryMarshal.CreateReadOnlySpan(ref _element0, 10);  
+    }
+}
+```
+
+The actual naming conventions for the type and its members are TBD. The framework will likely include a set of predefined "buffer" types that cover
+a limited set of buffer sizes. When a predefined type doesn't exist, compiler will synthesize it in the module being built. Names of the generated types
+will be "speakable" in order to support consumption from other languages. 
+
+A code generated for an access like: 
+``` C#
+public partial class C
+{
+    void M1(int val)
+    {
+        buffer1[1] = val;
+    }
+
+    int M2()
+    {
+        return buffer2[1];
+    }
+}
+```
+
+will be equivalent to:
+``` C#
+public partial class C
+{
+    void M1(int val)
+    {
+        buffer.AsSpan()[1] = val;
+    }
+
+    int M2()
+    {
+        return buffer2.AsReadOnlySpan()[1];
+    }
+}
+```
+
+#### Metadata import
+
+When compiler imports a field declaration of type *T* and the following conditions are all met:
+- *T* is a struct type decorated with the ```InlineArray``` attribute, and
+- The first instance field declared within *T* has type *F*, and
+- There is a ```public System.Span<F> AsSpan()``` within *T*, and 
+- There is a ```public readonly System.ReadOnlySpan<T> AsReadOnlySpan()``` or ```public System.ReadOnlySpan<T> AsReadOnlySpan()``` within *T*. 
+
+the field will be treated as C# fixed-size buffer with element type *F*. Otherwise, the field will be treated as a regular field of type *T*.
+
+
+## Detailed Design (Option 2)
```

Comment:
>(Option 2) [](http://example.com/codeflow?start=19&length=10)

Are the differences between the two options limited to element access? If so, consider removing the common items from the second option, so the differences are clear.

Reply:
> Are the differences between the two options limited to element access?

No they are not limited to just to element access. The first option doesn't have special treatment of fixed-size buffer types, there is a difference in field type from the language point of view, any occurrences of fixed-size buffer types are  affected (not just fields), etc.